stable_baselines/common/callbacks.py

```diff
@@ -0,0 +1,226 @@
+        if self.n_calls % self.save_freq == 0:
+            path = os.path.join(self.save_path, '{}_{}_steps'.format(self.name_prefix, self.num_timesteps))
+            self.model.save(path)
+            if self.verbose > 1:
+                print("Saving model checkpoint to {}".format(path))
+        return True
+
+
+class LambdaCallback(BaseCallback):
+    """
+    :param on_training_start: (callable)
+    :param on_step: (callable)
+    :param on_training_end: (callable)
+    :param verbose: (int)
+    """
+    def __init__(self, on_training_start=None, on_step=None, on_training_end=None, verbose=0):
+        super(LambdaCallback, self).__init__(verbose)
+        if on_training_start is not None:
+            self.on_training_start = on_training_start
+        if on_step is not None:
+            self._on_step = on_step
+        else:
+            self._on_step = lambda _locals, _globals: True
+        if on_training_end is not None:
+            self.on_training_end = on_training_end
+
+    def on_step(self, locals_: dict, globals_: dict) -> bool:
+        return self._on_step(locals_, globals_)
+
+
+class EvalCallback(BaseCallback):
+    """
+    Callback for evaluating an agent.
+
+    :param eval_env: (Union[gym.Env, VecEnv]) The environment used for initialization
+    :param n_eval_episodes: (int) The number of episodes to test the agent
+    :param eval_freq: (int) Evaluate the agent every eval_freq call of the callback.
+    :param log_path: (str) Path to a log file (.npz) where the evaluations
+        will be saved. It will be updated at each evaluation.
+    :param best_model_save_path: (str) Path to a folder where the best model
+        according to performance on the eval env will be saved.
+    :param deterministic: (bool) Whether the evaluation should
+        use a stochastic or deterministic actions.
+    :param verbose: (int)
+    """
+    def __init__(self, eval_env: Union[gym.Env, VecEnv],
+                 n_eval_episodes=5, eval_freq=10000, log_path=None,
+                 best_model_save_path=None,
+                 deterministic=True, verbose=1):
+        super(EvalCallback, self).__init__(verbose=verbose)
+        self.n_eval_episodes = n_eval_episodes
+        self.eval_freq = eval_freq
+        self.best_mean_reward = -np.inf
+        self.deterministic = deterministic
+        if isinstance(eval_env, VecEnv):
+            assert eval_env.num_envs == 1, "You must pass only one environment for evaluation"
+
+        self.eval_env = eval_env
+        self.best_model_save_path = best_model_save_path
+        self.log_path = log_path
+        self.evaluations_results = []
+        self.evaluations_timesteps = []
+
+    def init_callback(self, model):
+        super(EvalCallback, self).init_callback(model)
+        # Does not work when eval_env is a gym.Env and training_env is a VecEnv
+        # assert type(self.training_env) is type(self.eval_env), ("training and eval env are not of the same type",
```

Comment:
Why commented out?

Reply:
cf comment, you can have some false positive, maybe a warning instead?

Reply:
A warning would be good, at least to give hints in case people are running into weird errors.